Implement I/O methods for gammapy.maps

woodmd opened the pull request:
This PR fleshes out more of the FITS I/O related functionality in `gammapy.maps` and adds some new features to `HpxMapND`:
* Implement `read` and `write` methods for `WcsMapND` and `HpxMapND`.  
* Maps can be written to a sparse format by calling `write` with `sparse=True`.  This writes a BINTABLE with one row per pixel.  The format for HPX maps is documented on the [gamma-astro-data-formats page](http://gamma-astro-data-formats.readthedocs.io/en/latest/skymaps/healpix/index.html).  WCS maps use a similar convention which I will try to document in the next few weeks.  `WcsMapND.read` and `HpxMapND.read` can instantiate a map from a FITS HDU in either the sparse or non-sparse format.
* Support for instantiating `HpxMapND` with a multi-resolution geometry (i.e. different pixel size in each image plane).  This is implemented by allocating an array with an image dimension with size equal to the largest image plane in the map and filling array elements outside the geometry in a given image plane with guard values.  It's not particularly memory efficient but it should be useful for testing multi-resolution maps until the sparse map classes are fully implemented.

cdeil commented on gammapy/maps/base.py:
```diff
@@ -101,6 +102,57 @@ def create(cls, **kwargs):
+        """
+        hdulist = self.to_hdulist(**kwargs)
+        overwrite = kwargs.get('overwrite', True)
+        hdulist.writeto(filename, overwrite=overwrite)
```
There's a test fail on Windows:
```
 WindowsError: [Error 32] The process cannot access the file because it is being used by another process
```
https://ci.appveyor.com/project/cdeil/gammapy/build/1.0.2536/job/ej6xbtb9bpu7j7g1#L1570

Looking at http://docs.astropy.org/en/stable/_modules/astropy/io/fits/hdu/hdulist.html#HDUList.writeto I see that "close" is already called.
So to me, your I/O calls look right and I had hoped that this way every `read` and `write` call would never leave files open. But apparently that's not the case on Windows? (Or there's an issue where different pytest tests are running on the same test files in parallel. I don't think that's the case, but I'm not sure).

I'm not really sure what's going on / how to resolve this. Maybe mark the test as xfail on Windows so that this can go in, and then investigate / fix the Windows issue later, possibly asking Windows / Astropy people for help.

woodmd commented:
Thanks for taking a look.  I saw that error and was also stumped by it.  Unfortunately it's not easy for me to debug not having access to a windows machine.  Is there a way to mark a test with xfail for a specific platform (i.e. windows)?

woodmd commented:
I went ahead and took the opportunity to include some more updates with this PR.  Some of the additional changes:

- Added framework to support writing files in formats that are backwards compatible with existing software like the Fermi ScienceTools.  When you write a file you can set the `conv` parameter to enforce that the file is written in the appropriate format.  This mostly comes down to restricting the name of the BANDS table.  Ideally in the future all relevant software will be updated to support the formats documented in GADF but in the meantime we will probably need to keep some mechanism like this.  Note that we have started to implement partial support for some of the GADF conventions in the Fermi Science Tools.
- Implemented `reproject` method for WCS-based maps.  There are a few cases that are still missing (e.g. 4D maps where the non-spatial axes are not aligned).  One issue that arose was the lack of a method for wrapping a reprojection at the edge of an all-sky CAR map (e.g. like the fermi IEM).  I opened astrofrog/reproject#135 to request this feature.  For the moment I wrote some functions in `gammapy.maps.reproject` which can handle this but eventually it would be nice if this was included in the reproject package. 
- Added methods for iterating over the contents of a map.
- Various improvements to docstrings and sphinx documentation pages.

The original test failure on Windows / Appveyor also seems to be resolved now (although I'm not sure what I did to fix it).
